src/txkube/_authentication.py

```diff
@@ -173,3 +178,66 @@ def authenticating_agent(reactor, base_url, client_cert, client_key, ca_cert):
         },
     )
     return Agent(reactor, contextFactory=policy)
+
+
+
+@implementer(IAgent)
+class HeaderInjectingAgent(PClass):
+    """
+    An ``IAgent`` which adds some headers to every request it makes.
+
+    :ivar Headers _to_inject: The headers to add.
+    @ivar IAgent _agent: The agent to use to issue requests.
+    """
+    _to_inject = field(mandatory=True)
+    _agent = field(mandatory=True)
+
+    def request(self, method, url, headers=None, bodyProducer=None):
+        """
+        Issue a request with some extra headers.
+
+        :see: ``twisted.web.iweb.IAgent.request``
+        """
+        if headers is None:
+            headers = Headers()
+        else:
+            headers = headers.copy()
+        for k, vs in self._to_inject.getAllRawHeaders():
+            headers.setRawHeaders(k, vs)
+        return self._agent.request(method, url, headers, bodyProducer)
+
+
+
+def authenticate_with_serviceaccount(reactor, **kw):
+    """
+    Create an ``IAgent`` which can issue authenticated requests to a
+    particular Kubernetes server using a service account token.
+
+    :param reactor: The reactor with which to configure the resulting agent.
+
+    :param bytes path: The location of the service account directory.  The
+        default should work fine for normal use within a container.
+
+    :return IAgent: An agent which will authenticate itself to a particular
+        Kubernetes server and which will verify that server or refuse to
+        interact with it.
+    """
+    config = KubeConfig.from_service_account(**kw)
+
+    token = config.user["token"]
+    base_url = URL.fromText(config.cluster["server"].decode("ascii"))
+    [ca_cert] = pem.parse(config.cluster["certificate-authority"].bytes())
```

Comment:
alongside that, it might be worth adding some defenses here to deal with malformed CA data in the service account file (unless there is a guarantee elsewhere that the provider will not screw this up)